### PR TITLE
Feature/subscription filters

### PR DESCRIPTION
**Pull request by mschristensen**
We are creating a new method called getDerived so we can create derived/synthetic channels. Presently, we only have filter as an option. The filter option lets us implement filtered subscription on a channel.
DR can be found [here](https://ably.atlassian.net/wiki/spaces/product/pages/2560098305/DR2+Filtered+Subscription+SDK+Implementation+Decision)
Ticket: [DSC-54](https://ably.atlassian.net/browse/DSC-54)

Recreating this PR originally merged in https://github.com/ably/ably-js/pull/1185 and reverted in https://github.com/ably/ably-js/pull/1259 to apply these changes on an integration branch.

[DSC-54]: https://ably.atlassian.net/browse/DSC-54?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: test/realtime/message.test.js [hunk 1]
```diff
@@ -1183,5 +1183,145 @@ define(['ably', 'shared_helper', 'async', 'chai'], function (Ably, helper, async
...
+                }
+              } catch (err) {
+                closeAndFinish(done, realtime, err);
+                return;
+              }
+              closeAndFinish(done, realtime);
+            });
+          });
+          var restChannel = rest.channels.get('chan');
+          restChannel.publish(testData);
```
Comment (mschristensen) [inline test/realtime/message.test.js:1319]: @mohyour it looks like there still might be a race here, we are publishing when we are connected but we are not guaranteed to publish after we have attached to the filtered & unfiltered channel. If messages are published before this attachment succeeds, we will lose messages. Instead, we should wait for the attachments before publishing.

File: test/realtime/message.test.js [hunk 2]
```diff
@@ -1183,5 +1183,145 @@ define(['ably', 'shared_helper', 'async', 'chai'], function (Ably, helper, async
...
+                  expect(unFilteredMessages[i].data).to.equal(testData[i].data, 'Unexpected data received');
+                }
+              } catch (err) {
+                closeAndFinish(done, realtime, err);
+                return;
+              }
+              closeAndFinish(done, realtime);
+            });
+            var restChannel = rest.channels.get('chan');
+            restChannel.publish(testData);
```
Comment (mschristensen) [inline test/realtime/message.test.js:1318]: Hmm, almost there. While this fixes the issue for the unfiltered channel attachment, we still have a race condition where the test might fail if the filtered channel has not yet attached before we start publishing messages. We need to ensure both attachments have succeeded before we publish :)

**Review by owenpearson — COMMENTED** [review]
LGTM, thanks!

**Review by owenpearson — APPROVED** [review]
LGTM, thanks!

**Review by owenpearson — APPROVED** [review]
LGTM, thanks!

**Review by owenpearson — APPROVED** [review]
LGTM, thanks!

**Review by owenpearson — APPROVED** [review]
LGTM, thanks!

**Pull request merged**